main.go

```diff
@@ -1,126 +1,131 @@
+	}, []string{
+		SensorID,
+		SensorLocation,
+	})
+
+	prometheus.MustRegister(temperature)
+	prometheus.MustRegister(humidity)
 
 	http.Handle("/metrics", promhttp.Handler())
 
-	go receive(device)
-
-	log.Printf("Serving metrics at '%v/metrics'", *listenAddr)
-	log.Fatal(http.ListenAndServe(*listenAddr, nil))
-}
-
-func receive(device *string) {
-	a := &arduino.Device{}
-
-	err := a.Open(*device)
+	dev, err := OpenDevice(*device)
 	if err != nil {
 		log.Fatalf("Could not open '%v'", *device)
 	}
+	defer dev.Close()
 
-	// wait until the Arduino is ready to accept commands
-	err = a.Read(Ready{})
+	err = dev.Reset()
 	if err != nil {
-		log.Fatalf("Device is not ready to take commands: %s", err)
+		log.Fatalf("Could not reset '%v'", *device)
 	}
 
+	go receive(dev)
+
+	log.Printf("Serving metrics at '%v/metrics'", *listenAddr)
+	log.Fatal(http.ListenAndServe(*listenAddr, nil))
+}
+
+func receive(a *Device) {
 	// tell the Arduino to start receiving signals
-	err = a.Write(arduino.ReceiveCmd)
+	err := a.Write(ReceiveCmd)
 	if err != nil {
-		log.Fatalf("Could not write to '%v'", *device)
+		log.Fatalf("Could not write to '%v'", a)
 	}
+	log.Println("Write", ReceiveCmd)
 
+	ctx := context.Background()
 	// read and decode received signals forever
-	err = a.Read(DecodedSignal{})
+	err = a.Process(ctx, DecodedSignal)
 	if err != nil {
 		log.Println(err)
 	}
 }
 
-// Ready implements a Processor that waits and returns
-// as soon as the Arduino is ready to accept commands.
-type Ready struct{}
-
-// Process reads from the Arduino until it returns "ready".
-func (Ready) Process(s string) bool {
-	return strings.Contains(s, arduino.Ready)
-}
-
-// DecodedSignal implements a Processor that
-// decodes received raw signals and sets the Prometheus Gauge
-// based on the decoded values.
-type DecodedSignal struct{}
-
 // Process decodes a compressed signal read from the Arduino
 // by trying all currently supported protocols.
-func (DecodedSignal) Process(line string) bool {
-	if strings.HasPrefix(line, arduino.ReceivePrefix) {
-		trimmed := strings.TrimPrefix(line, arduino.ReceivePrefix)
+func DecodedSignal(line string) (stop bool) {
+	stop = false
+
+	if strings.HasPrefix(line, ReceivePrefix) {
+		trimmed := strings.TrimPrefix(line, ReceivePrefix)
 
-		p, err := pulse.Prepare(trimmed)
+		p, err := PreparePulse(trimmed)
 		if err != nil {
 			log.Println(err)
-			return true
+			return
 		}
 
-		result, err := pulse.Decode(p)
+		device, result, err := DecodePulse(p)
 		if err != nil {
 			log.Println(err)
-			return true
+			return
 		}
 
-		if result != nil {
-			m := result.(*pulse.GTWT01Result)
-			log.Printf("%+v\n", *m)
-
-			if t, ok := temperature[m.ID]; ok {
-				t.Set(m.Temperature)
-			}
-			if h, ok := humidity[m.ID]; ok {
-				h.Set(float64(m.Humidity))
+		switch device {
+		case GT_WT_01:
+			m := result.(*GTWT01Result)
+			log.Printf("%v: %+v\n", device, *m)
+			if loc, ok := sensorLocations[m.Name]; !ok && loc != "" {
```

Comment:
Check if there is a sensor id is put in the map and the `value` is not empty string